core/valor_core/managers.py

```diff
@@ -590,3 +598,229 @@ def evaluate(self):
+                    or pred.datum.metadata != pred2.datum.metadata
+                ):
+                    raise MismatchingTextGenerationDataError(
+                        f"Two predictions with the same datum_uid {pred.datum.uid} have different datum text or metadata."
+                    )
+
+        eval = text_generation.evaluate_text_generation(
+            predictions=predictions,
+            metrics_to_return=self.metrics_to_return,
+            llm_api_params=self.llm_api_params,
+        )
+
+        for pred in predictions:
+            self.joint_df = pd.concat(
+                [
+                    self.joint_df,
+                    pd.DataFrame(
+                        [
+                            {
+                                "datum_uid": pred.datum.uid,
+                                "datum_text": pred.datum.text,
+                                "datum_metadata": pred.datum.metadata,
+                                "prediction_text": annotation.text,
+                                "prediction_context_list": annotation.context_list,
+                            }
+                            for annotation in pred.annotations
+                        ],
+                    ),
+                ],
+                ignore_index=True,
+            )
+
+        # Conditions for matching computed metrics to the correct row.
+        def conditions(
+            row, datum_uid, prediction_text=None, prediction_context_list=None
+        ):
+            if prediction_text is not None:
+                if prediction_context_list is not None:
+                    return (
+                        row["datum_uid"] == datum_uid
+                        and row["prediction_text"] == prediction_text
+                        and row["prediction_context_list"]
+                        == prediction_context_list
+                    )
+                else:
+                    return (
+                        row["datum_uid"] == datum_uid
+                        and row["prediction_text"] == prediction_text
+                    )
+            else:
+                return (
+                    row["datum_uid"] == datum_uid
+                    and row["prediction_context_list"]
+                    == prediction_context_list
+                )
+
+        for m in eval.metrics:
+            metric_name = m["type"]
+            value = m["value"]
+            datum_uid = m["parameters"]["datum_uid"]
+            prediction_text = m["parameters"].get("prediction", None)
+            prediction_context_list = m["parameters"].get("context_list", None)
+
+            # Set the metric value of the correct row.
+            conditional_series = self.joint_df.apply(
+                conditions,
+                axis=1,
+                args=(datum_uid, prediction_text, prediction_context_list),
+            )
+            self.joint_df.loc[conditional_series, metric_name] = value
```

Comment:
None of the code in the streaming manager prevents the addition of duplicate predictions. If this function is called with a prediction annotation that matches the exact same Datum, prediction.text and prediction.context_list, then even though that data is already in the joint_df, the metrics will be computed again and a new row will be added to the joint_df. Further, because of the way that I add the metric results to the joint_df, the second metric results will also overwrite the first metric results for the first row in the joint_df. This is potentially an issue that an old row in the dataframe could get updated, however this is not a high priority issue. The most likely best fix for this issue would be to prevent the addition of identical prediction annotations, but that would take some work to setup and test.